spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -127,19 +128,48 @@ private[cassandra] class CassandraSourceRelation(
   /** Construct Cql clause and retrieve the values from filter */
   private def filterToCqlAndValue(filter: Any): (String, Seq[Any]) = {
     filter match {
-      case sources.EqualTo(attribute, value)            => (s"${quote(attribute)} = ?", Seq(value))
-      case sources.LessThan(attribute, value)           => (s"${quote(attribute)} < ?", Seq(value))
-      case sources.LessThanOrEqual(attribute, value)    => (s"${quote(attribute)} <= ?", Seq(value))
-      case sources.GreaterThan(attribute, value)        => (s"${quote(attribute)} > ?", Seq(value))
-      case sources.GreaterThanOrEqual(attribute, value) => (s"${quote(attribute)} >= ?", Seq(value))
+      case sources.EqualTo(attribute, value)            => (s"${quote(attribute)} = ?", Seq(toCqlValue(attribute, value)))
+      case sources.LessThan(attribute, value)           => (s"${quote(attribute)} < ?", Seq(toCqlValue(attribute, value)))
+      case sources.LessThanOrEqual(attribute, value)    => (s"${quote(attribute)} <= ?", Seq(toCqlValue(attribute, value)))
+      case sources.GreaterThan(attribute, value)        => (s"${quote(attribute)} > ?", Seq(toCqlValue(attribute, value)))
+      case sources.GreaterThanOrEqual(attribute, value) => (s"${quote(attribute)} >= ?", Seq(toCqlValue(attribute, value)))
       case sources.In(attribute, values)                 =>
-        (quote(attribute) + " IN " + values.map(_ => "?").mkString("(", ", ", ")"), values.toSeq)
+        (quote(attribute) + " IN " + values.map(_ => "?").mkString("(", ", ", ")"), toCqlValues(attribute, values))
       case _ =>
         throw new UnsupportedOperationException(
           s"It's not a valid filter $filter to be pushed down, only >, <, >=, <= and In are allowed.")
     }
   }
 
+  private def toCqlValues(columnName: String, values: Array[Any]): Seq[Any] = {
+    if (VarIntColumn(columnName)) {
+      values.map(toBigInt).toSeq
+    } else {
+      values.toSeq
+    }
+  }
+
+  /** If column is VarInt column, convert data to BigInteger */
+  private def toCqlValue(columnName: String, value: Any): Any = {
+    if (VarIntColumn(columnName)) {
+      toBigInt(value)
+    } else {
+      value
+    }
+  }
+
+  private def VarIntColumn(columnName: String): Boolean = {
+    tableDef.columnByName(columnName).columnType == VarIntType
+  }
+
+  /** Convert decimal to BigInteger */
+  private def toBigInt(value: Any): Any = {
```

Comment:
why not lock the input of this function to decimal? Also think about naming decimalToBigInt? Then error out if called on non-bigint column

I think the best solution is to make this function and lock input to the type decimal then in toCqlValue do a match on type so the compiler will be happy